tests/nn/data_parallel/test_sharded_ddp_pytorch_parity.py

```diff
@@ -64,132 +65,134 @@ def run_ddp_parity(
+        broadcast_buffers=True,
+        reduce_buffer_size=reduce_buffer_size,
+        reduce_fp16=fp16_reduction,
+    )
+
+    ddp_model_single = copy.deepcopy(model)
+    ddp_optimizer = torch.optim.SGD(ddp_model_single.parameters(), lr=1e-4, momentum=0.99)
+    ddp_model = DDP(ddp_model_single, device_ids=[rank], broadcast_buffers=True, find_unused_parameters=True)
+
+    if fp16_reduction:
+        from dist.algorithms.ddp_com_hooks.default_hooks import fp16_compress_hook
+
+        ddp_model.register_comm_hook(state=None, hook=fp16_compress_hook)  # type: ignore
+
+    ddp_scaler = TorchGradScaler() if amp else None
+    sharded_scaler = ShardedGradScaler() if amp else None
+
+    # The model should be synchronized in between the ranks at construction time, check that
+    check_same_model_params(sharded_ddp_model, ddp_model)
+
+    # Typical training loop, check that we get the exact same results as DDP
+    for i in range(NUMBER_BATCHS):
+        input_tensor = torch.rand((BATCH_SIZE, 2)).to(device)
+
+        def ddp_closure(input_tensor=input_tensor):
+            return closure(ddp_model, ddp_scaler, input_tensor, grad_accumulation)
+
+        def sharded_closure(input_tensor=input_tensor):
+            return closure(
+                sharded_ddp_model, sharded_scaler, input_tensor, grad_accumulation, _manual_reduction=manual_reduction,
+            )
+
+        # Step/scale both
+        for _scaler, _closure, _optimizer in (
+            (ddp_scaler, ddp_closure, ddp_optimizer),
+            (sharded_scaler, sharded_closure, sharded_optimizer),
+        ):
+            if _scaler is not None:
+                _ = _closure(input_tensor)
+                _scaler.step(_optimizer)
+                _scaler.update()
+            else:
+                _optimizer.step(_closure())
+
+        check_same_model_params(sharded_ddp_model, ddp_model, f"Rank: {rank} - Step {i} broke")
+
+        # Check that the two grad norm are equivalent
+        # NOTE: The grads can occasionally be NaNs, the scaler will skip the step in that case
+        # This is not ShardedDDP specific. If the grads are not NaN for DDP then they should also
+        # be valid for ShardedDDP
+        # NOTE: DDP does not handle parameters trainability being changed after the fact, see
+        # https://github.com/pytorch/pytorch/blob/5781aec74ef00284e0262817a649278c2e8072bf/torch/nn/parallel/distributed.py#L471
+        if clip_grad_norm and not change_train_graph:
+            total_norm = torch.nn.utils.clip_grad_norm_(ddp_model.parameters(), 0.3, norm_type=2.0)  # type: ignore
+            if not torch.isnan(total_norm):
+                oss_total_norm = sharded_optimizer.clip_grad_norm(0.3, norm_type=2.0)
+                allclose = torch.allclose(oss_total_norm, total_norm, atol=1e-2 if amp else 1e-8)
+
+                if not allclose:
+                    # Debug helper if this unit test does not pass, compare the gradients in between DDP and ShardedDDP
```

Comment:
adding a debug printout in case this test breaks in the future, makes it easier to see if and where the reduction broke